File: longest-line-of-consecutive-one/grandyang_dp.java
//grandyang dp idea, shaun java code, runtime 16ms, faster than 63%
class Solution {
    public int longestLine(int[][] M) {
        if (M == null || M.length == 0) return 0;
        int m = M.length, n = M[0].length, res = 0;
        //dp[i][j][k]表示从开头遍历到数字nums[i][j]为止，第k种情况的连续1的个数
        int [][][] dp = new int[m][n][4];
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (M[i][j] == 0) continue;
                //M[i][j] == 1, path len 1, not quite understood the following line 
                for (int k = 0; k < 4; ++k) dp[i][j][k] = 1;
                
                if (j > 0) dp[i][j][0] += dp[i][j - 1][0]; // horizonal
                if (i > 0) dp[i][j][1] += dp[i - 1][j][1]; // vertical
                if (i > 0 && j < n - 1) dp[i][j][2] += dp[i - 1][j + 1][2]; // diagonal
                if (i > 0 && j > 0) dp[i][j][3] += dp[i - 1][j - 1][3]; // anti-diagonal
                for (int k = 0; k < 4; k++)
                    res = Math.max(res, dp[i][j][k]);
            }
        }
        return res;
    }
}

File: split-array-largest-sum/grandyang_dp.cpp
//grandyang dp idea, runtime 48ms, faster than 29%
//to understand, Feb 2 2019
//not easy to come up with dp[m][n], m -- number of arrays, n: array from 0 to n-1
class Solution {
public:
    int splitArray(vector<int>& nums, int m) {
        int n = nums.size();
        vector<int> sums(n + 1, 0);
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, INT_MAX));
        dp[0][0] = 0;
        for (int i = 1; i <= n; ++i) {
            sums[i] = sums[i - 1] + nums[i - 1];
        }
        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= n; ++j) {
                for (int k = i - 1; k < j; ++k) {
                    int val = max(dp[i - 1][k], sums[j] - sums[k]);
                    dp[i][j] = min(dp[i][j], val);
                }
            }
        }
        return dp[m][n];
    }
};
